### PR TITLE
[25.0 backport] update to Go 1.21.12

### PR DESCRIPTION
**Pull request by austinvazquez**
**- What I did**

Backports:
1. #47405
1. #47902
1. #48120 - + set ALPINE_VERSION to 3.19 (golang 3.18 images are no longer available)
1. #48121

**- How I did it**
```
git cherry-pick -xsS 8761bffcafd39a99a4fdadf919f7a8d15c46aa2a
git cherry-pick -xsS 91e2c29865184924451ec03f675c0bfe91b63ac2
git cherry-pick -xsS 4d1d7c3ebe02fe9762dd35411dd4421afca10c99
git cherry-pick -xsS 837289ba626df49d04123885e9e721039f8842e1
```

\* Resolve conflicts along the way

**- How to verify it**
CI should be successful

**- Description for the changelog**

```markdown changelog
Update toolchain to Go 1.21.12
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by austinvazquez** [comment]
Hey folks, I think this one is ready for another pair of eyes. Would it be possible to get a rerun for cross-compilation of linux/arm/v5? It looks like the job failed but without logs so possibly just a job launch failure. Also for snapshotter unit tests could it be related to #42484?

Thanks again for continuing to accept support for 25.0; it has really helped us on the distribution side where upgrades to modern versions are expensive. Hoping this will get us over the line for 25.0.6.

File: .github/workflows/buildkit.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ on:
   pull_request:
 
 env:
-  GO_VERSION: "1.21.10"
+  GO_VERSION: "1.21.12"
```
Comment (corhere) [inline .github/workflows/buildkit.yml:16]:
```suggestion
  GO_VERSION: "1.21.12"
  ALPINE_VERSION: "3.19"
```

**Review by corhere — CHANGES_REQUESTED** [review]
Bumping the Buildkit major version is a breaking change: it makes v25 into a franken-v26. We cannot upgrade buildkit to v0.13 in the v25 maintenance branch.

> The buildkit update is needed for 25.0 because alpine1.18 images are no longer supported for golang v1.21.11+.

Buildkit does not need to be updated to resolve that. Its `hack/test` script [respects an `ALPINE_VERSION` env var in addition to the `GO_VERSION` env var already in use.](https://github.com/moby/buildkit/blob/dbed8c0a42162d1f3e9ec89038d95575e78f8759/hack/test#L6-L7)

**Review by corhere — APPROVED** [review]
LGTM; thanks!

File: .github/workflows/buildkit.yml (hunk 1 above)
Comment (austinvazquez) [reply .github/workflows/buildkit.yml:16]: +1, thanks @corhere for the clarification!

**Pull request merged**